[CLOUDS-4012] Fix service tag setting in lambda forwarder

ge0Aja opened the pull request:
### What does this PR do?
Fixing a bug related to service tag setting in the lambda forwarder

This PR fixes the precedence of setting the service tag from multiple sources and removes the redundancy when having multiple values for the service tag from these sources.

The order of priority for setting the service tag is as follows:

1. `message` field in the log if it has a custom tags field set  i.e. `message.ddtags`
2. `ddtags` env var in the lambda logs forwarder 
3. LogGroup tags cache 
4. Lambda tags cache 
5. `ddsource` if non of the above is set
 
Previously the forwarder appended all the values set for the service from all sources in the custom tags field which created a confusion between the actual value of the service key in the log and the appended values in the custom tags field. 
This PR removes the redundant values based on the above priorities and ensures that the service key is set according to the value of the highest priority.
 
### Motivation
https://datadoghq.atlassian.net/browse/CLOUDS-4012

### Testing Guidelines

<!--- How did you test this pull request? --->

### Additional Notes

<!--- Anything else we should know when reviewing? --->

### Types of changes

- [x] Bug fix
- [ ] New feature
- [ ] Breaking change
- [ ] Misc (docs, refactoring, dependency upgrade, etc.)

### Check all that apply

- [x] This PR's description is comprehensive
- [ ] This PR contains breaking changes that are documented in the description
- [ ] This PR introduces new APIs or parameters that are documented and unlikely to change in the foreseeable future
- [ ] This PR impacts documentation, and it has been updated (or a ticket has been logged)
- [x] This PR's changes are covered by the automated tests
- [ ] This PR collects user input/sensitive content into Datadog
- [x] This PR passes the integration tests (ask a Datadog member to run the tests)
- [x] This PR passes the unit tests 
- [x] This PR passes the installation tests (ask a Datadog member to run the tests)

ge0Aja committed:
[CLOUDS-4012] Fix service tag setting in lambda forwarder

If the service tag in set in dd_tags use it, otherwise use the one coming from the logGroup
If none of the above, check if the tag is set on the lambda forwarder and use it, otherwise use the function name.
Finally, check if the service tag is set in the application level and override

ge0Aja committed:
Fix Unit tests, Integration tests, python formatting

tyrcho commented on aws/logs_monitoring/parsing.py:
```diff
@@ -242,15 +242,21 @@ def s3_handler(event, context, metadata):
 
-def get_service_from_tags(metadata):
-    # Get service from dd_custom_tags if it exists
+def get_service_from_tags_and_remove_duplicates(metadata):
```
you also need to rename the call from test_parsing.py

ge0Aja committed:
Fix Formatting on test_parsing.py

ge0Aja committed:
Fix a bug in setting the service tag from the message

ALso:
	- Update Itests
	- Add __init__.py to create a package for running utests inside the IDE

ge0Aja committed:
Add UTests

ge0Aja committed:
Update integrtion test files to match the new order of dd_tags keys

ge0Aja committed:
Make tags output order consistent when extracting message tags

ge0Aja committed:
Fix formatting

tyrcho commented on aws/logs_monitoring/lambda_function.py:
```diff
@@ -257,14 +257,24 @@ def add_metadata_to_lambda_log(event):
+            tags.append(service_tag)
+            event[DD_SERVICE] = service_tag.split(":")[1]
+    else:
+        # remove the service tag from the cusotm lambda tags if it exists
```
```suggestion
        # remove the service tag from the custom lambda tags if it exists
```

tyrcho commented on aws/logs_monitoring/lambda_function.py:
```diff
@@ -319,7 +329,54 @@ def extract_ddtags_from_message(event):
+            event[DD_CUSTOM_TAGS], extracted_ddtags
+        )
+
+        ## Extract service tag from message.ddtags if exists
```
```suggestion
        # Extract service tag from message.ddtags if exists
```

tyrcho commented on aws/logs_monitoring/tests/test_lambda_function.py:
```diff
@@ -231,5 +313,15 @@ def test_extract_trace_payload_valid_trace(self):
 
+class TestMergeMessageTags(unittest.TestCase):
+    def test_merge_custom_and_application_tags(self):
+        message_tags = "key0:value0,key1:value1"
```
can't we have repetition in tags here ? like `message_tags = "key0:value0,key1:value1,key1:value11"`

I'm wondering what is the expected / actual behavior with that

ge0Aja replied:
Not sure on this, but looking back at the code. During parsing, it doesn't seem that we account for multiple tags with the same key but a different value.

tyrcho commented on aws/logs_monitoring/lambda_function.py:
```diff
@@ -319,7 +329,54 @@ def extract_ddtags_from_message(event):
+            )
+
+
+def merge_custom_and_application_tags(custom_tags, application_tags):
```
Could we see this function as simpler / more generic ?
- signature could be `merge_tags(specific_tags, default_tags)`
- comment would only insist on the key feature imo : specific overrides defaut

The rest of the comment could move to the place where it's used at the moment

My gut feeling is that this coud probably be reused elsewhere.

tyrcho replied:
I'm also wondering if we also want another version of this which appends tags instead of replacing

 merge(k1:v1, k1:v2) => "k1:v1,k1:v2"

ge0Aja replied:
For the append version we can just f"{custom_tags},{application_tags}" as it was before. 
I don't think we need a separate method for that

tyrcho commented on aws/logs_monitoring/lambda_cache.py:
```diff
@@ -78,5 +78,12 @@ def get(self, key):
             logger.debug("Local cache expired, fetching cache from S3")
             self._refresh()
 
+        if not self.should_fetch_tags():
```
could we do this test before the refresh ?

tyrcho commented on aws/logs_monitoring/lambda_cache.py:
```diff
@@ -78,5 +78,12 @@ def get(self, key):
+                "not set to true"
+            )
+            return []
         function_tags = self.tags_by_id.get(key, [])
```
we can merge with next line

tyrcho commented on aws/logs_monitoring/lambda_function.py:
```diff
@@ -257,14 +257,24 @@ def add_metadata_to_lambda_log(event):
     custom_lambda_tags = get_enriched_lambda_log_tags(event)
 
-    # Set the `service` tag and metadata field. If the Lambda function is
+    # If not set during parsing or set with a default value
```
I have trouble understanding this comment. Maybe with extra indenting + punctuation at the end of this line ?

tyrcho commented on aws/logs_monitoring/lambda_function.py:
```diff
@@ -257,14 +257,24 @@ def add_metadata_to_lambda_log(event):
     # Get custom tags of the Lambda function
     custom_lambda_tags = get_enriched_lambda_log_tags(event)
 
-    # Set the `service` tag and metadata field. If the Lambda function is
```
Would you add on the main comment (l232) the rule we defined with the precedence for service tag ?

ge0Aja replied:
Changed the wording here

ge0Aja committed:
Address some review comments / fix typos

ge0Aja committed:
Fix failing utest after function rename

ge0Aja committed:
Change tag merging behavior between message ddtags and upper level ddtags

Override the service tag if it exists.
Merge upper level ddtags and message ddtags in one string.

ge0Aja committed:
Fix Itests results order
